docs/cryptodoc/src/05_09_frodokem.rst

```diff
@@ -0,0 +1,298 @@
+
+   **Output:**
+
+   -  ``SK``, ``PK``: private and public key
+
+   **Steps:**
+
+   1. Generate new values ``s``, ``seed_se``, and ``z`` using ``rng``
+   2. ``seed_a = SHAKE(z, len_a)``
+   3. ``r = SHAKE(0x5F || seed_se, 32*n*n_bar)``
+   4. ``s_trans = sample_matrix(r[:n*n_bar - 1])``
+   5. ``e = sample_matrix(r[n*n_bar:])``
+   6. ``b = a*s + e``
+   7. | ``PK = {seed_a, pack(b)}``
+      | ``pkh = SHAKE(PK, len_sec)``
+      | ``SK = {s, seed_a, pack(b), s_trans, pkh}``
+
+   **Notes:**
+
+   - Computation of ``b = a*s + e`` is done by a specialised function that performs on-demand
+     expansion of ``seed_a`` into the desired row of the matrix :math:`A` of [FrodoKEM-ISO]_
+     and assumes getting the transpose ``s_trans`` of ``s`` as input to avoid transposition.
+   - The operation ``pack(b)`` is performed when accessing the serialized or raw key bits of
+     a key.
+   - The computation of ``pkh`` is performed in the constructor of ``FrodoKEM_PublicKeyInternal``,
+     an internal class used to represent ``PK``.
+   - The creation of a ``FrodoKEM_PublicKey`` is conducted using the
+     ``public_key`` method of the private key.
+
+..  _pubkey/frodokem/encapsulation:
+
+Key Encapsulation
+-----------------
+
+The FrodoKEM encapsulation procedure of Botan follows Section 8.2 of [FrodoKEM-ISO]_ and
+works as follows:
+
+.. admonition:: FrodoKEM Encapsulation
+
+   **Input:**
+
+   - ``PK = {seed_a, packed_b}``: public key
+   - ``rng``: random number generator
+
+   **Output:**
+
+   - ``encapsulated_key``: ciphertext of shared key
+   - ``shared_key``: plaintext shared key
+
+   **Steps:**
+
+   1. Generate new values ``u`` and ``salt`` using ``rng``
+   2.  ``seed_se || k = SHAKE(pkh || u || salt, len_se + len_sec)``
+   3. ``r = SHAKE(0x96 || seed_se, 16*(2*n_bar*n + n_bar*n_bar))``
+   4. ``s_p = sample_matrix(r[:n*n_bar - 1])``
+   5. ``e_p = sample_matrix(r[n*n_bar:2*n*n_bar - 1])``
+   6. ``b_p = s_p*a + e_p``
+   7. ``c_1 = pack(b_p)``
+   8. ``e_pp = sample_matrix(r[2*n*n_bar:])``
+   9.  ``v = s_p*b + e_pp``
+   10. ``c = v + encode(u)``
+   11. ``c_2 = pack(c)``
+   12. ``encapsulated_key = c_1 || c_2 || salt``
+   13. ``shared_key = SHAKE(encapsulated_key || k, len_sec)``
+
+   **Notes:**
+
+   - The computation of ``pkh`` is already performed in the constructor of ``FrodoKEM_PublicKeyInternal``,
+     an internal class used to represent ``PK``.
+   - ``b = unpack(packed_b)`` is already performed in the constructor of ``FrodoKEM_PublicKey``.
+   - The computation of ``s_p*a + e_p`` is done by a specialised function that performs on-demand
+     expansion of ``seed_a`` into the desired row of the matrix :math:`A` of [FrodoKEM-ISO]_.
+   - The computation of ``s_p*b + e_pp`` is done by a specialised function realising this sequence
+     of operations.
+
+..  _pubkey/frodokem/decapsulation:
+
+Key Decapsulation
+-----------------
+
+The FrodoKEM encapsulation procedure of Botan follows Section 8.3 of [FrodoKEM-ISO]_ and
```

Comment:
While copying/pasting the structure for our Classic McElieces crypto doc, I noticed this one:
```suggestion
Key Decapsulation
-----------------

The FrodoKEM decapsulation procedure of Botan follows Section 8.3 of [FrodoKEM-ISO]_ and
```